File: Backend/Reservaciones/reservaciones.js
let reservaciones = [];
let editandoId = null;

const form = document.getElementById("reservationForm");
const body = document.getElementById("reservationsBody");
const searchByName = document.getElementById("searchByName");

const nombreInput = document.getElementById("nombre");
const telefonoInput = document.getElementById("telefono");
const fechaInput = document.getElementById("fecha");
const horaInput = document.getElementById("hora");
const personasInput = document.getElementById("personas");
const estadoInput = document.getElementById("estado");
const notaInput = document.getElementById("nota");

const totalToday = document.getElementById("totalToday");
const upcomingCount = document.getElementById("upcomingCount");
const nextReservationLabel = document.getElementById("nextReservationLabel");
const nextReservationTime = document.getElementById("nextReservationTime");

const btnLimpiar = document.getElementById("btnLimpiar");
const btnBack = document.getElementById("btnBack");

function formatearFecha(fecha) {
    if (!fecha) return "—";
    return fecha;
}

function formatearHora(hora) {
    if (!hora) return "—";
    return hora.slice(0, 5);
}

function limpiarFormulario() {
    form.reset();
    editandoId = null;
}

async function cargarReservaciones() {
    try {
        const res = await fetch("/api/reservaciones");
        const data = await res.json();

        reservaciones = Array.isArray(data) ? data : [];
        renderTabla();
        renderCards();
    } catch (err) {
        console.error("Error al cargar reservaciones:", err);
        alert("No se pudieron cargar las reservaciones.");
    }
}

function renderTabla() {
    body.innerHTML = "";

    const filtro = searchByName.value.trim().toLowerCase();

    const filtradas = reservaciones.filter(r => {
        const nombreCompleto = `${r.nombre_cliente} ${r.apellido_cliente || ""}`.toLowerCase();
        return nombreCompleto.includes(filtro);
    });

    filtradas.forEach(r => {
        const tr = document.createElement("tr");

        tr.innerHTML = `
            <td>${formatearHora(r.hora)}</td>
            <td>${r.nombre_cliente} ${r.apellido_cliente || ""}</td>
            <td>${r.no_personas}</td>
            <td>${formatearFecha(r.fecha)}</td>
            <td>
            <span class="estado-badge estado-${r.estado.toLowerCase()}">
            ${r.estado}
            </span>
            </td> 
            <td>
                <button class="table-btn table-btn--small" onclick="editarReservacion(${r.id_reservacion})">Editar</button>
                <button class="table-btn table-btn--danger table-btn--small" onclick="eliminarReservacion(${r.id_reservacion})">Eliminar</button>
            </td>
        `;

        body.appendChild(tr);
    });
}

function renderCards() {
    const hoy = new Date().toISOString().split("T")[0];
    const ahora = new Date();

    const hoyReservaciones = reservaciones.filter(r => r.fecha === hoy);
    totalToday.textContent = hoyReservaciones.length;

    const proximas = hoyReservaciones.filter(r => {
        const fechaHora = new Date(`${r.fecha}T${r.hora}`);
        const diff = (fechaHora - ahora) / (1000 * 60);
        return diff >= 0 && diff <= 60;
    });

    upcomingCount.textContent = proximas.length;

    const futuras = reservaciones
        .map(r => ({ ...r, fechaHora: new Date(`${r.fecha}T${r.hora}`) }))
        .filter(r => r.fechaHora >= ahora)
        .sort((a, b) => a.fechaHora - b.fechaHora);

    if (futuras.length > 0) {
        nextReservationLabel.textContent = `${futuras[0].nombre_cliente} ${futuras[0].apellido_cliente || ""}`;
        nextReservationTime.textContent = `${futuras[0].fecha} • ${formatearHora(futuras[0].hora)}`;
    } else {
        nextReservationLabel.textContent = "Sin próximas";
        nextReservationTime.textContent = "—";
    }
}

form.addEventListener("submit", async function (e) {
    e.preventDefault();

    const payload = {
        no_empleado: parseInt(localStorage.getItem("NO_EMPLEADO")) || 1001,
        nombre_cliente: nombreInput.value.trim(),
        apellido_cliente: "",
        telefono: telefonoInput.value.trim(),
        fecha: fechaInput.value,
        hora: horaInput.value + ":00",
        no_personas: parseInt(personasInput.value),
        estado: estadoInput.value,
        comentarios: notaInput.value.trim()
    };

    if (!payload.nombre_cliente || !payload.telefono || !payload.fecha || !payload.hora || !payload.no_personas || !payload.estado) {
        alert("Completa todos los campos obligatorios.");
        return;
    }

    try {
        const url = editandoId ? `/api/reservaciones/${editandoId}` : "/api/reservaciones";
        const metodo = editandoId ? "PUT" : "POST";

        const res = await fetch(url, {
            method: metodo,
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(payload)
        });

        const data = await res.json();

        if (!res.ok) {
            alert(data.message || "Ocurrió un error.");
            return;
        }

        alert(data.message || "Operación realizada correctamente.");
        limpiarFormulario();
        await cargarReservaciones(); // auto update
    } catch (err) {
        console.error("Error al guardar reservación:", err);
        alert("Error al guardar la reservación.");
    }
});

function editarReservacion(id) {
    const r = reservaciones.find(x => x.id_reservacion === id);
    if (!r) return;

    editandoId = id;

    nombreInput.value = r.nombre_cliente || "";
    telefonoInput.value = r.telefono || "";
    fechaInput.value = r.fecha || "";
    horaInput.value = formatearHora(r.hora);
    personasInput.value = r.no_personas || "";
    estadoInput.value = r.estado || "Pendiente";
    notaInput.value = r.comentarios || "";
}

async function eliminarReservacion(id) {
    const confirmacion = confirm("¿Seguro que deseas eliminar esta reservación?");
    if (!confirmacion) return;

    try {
        const res = await fetch(`/api/reservaciones/${id}`, {
            method: "DELETE"
        });

        const data = await res.json();

        if (!res.ok) {
            alert(data.message || "No se pudo eliminar.");
            return;
        }

        alert(data.message || "Reservación eliminada correctamente.");
        await cargarReservaciones();
    } catch (err) {
        console.error("Error al eliminar reservación:", err);
        alert("No se pudo eliminar la reservación.");
    }
}

btnLimpiar.addEventListener("click", limpiarFormulario);
searchByName.addEventListener("input", renderTabla);

btnBack.addEventListener("click", () => {
    const rol = localStorage.getItem("ROL");

    if (rol === "GERENTE") window.location.href = "/gerente";
    else if (rol === "HOSTESS") window.location.href = "/hostess";
    else if (rol === "JEFEDEPISO") window.location.href = "/jefepiso";
    else window.location.href = "/";
});

cargarReservaciones();

File: Backend/Reservaciones/reservaciones.py
from flask import Blueprint, jsonify, render_template, request
from db import get_connection

reservaciones_bp = Blueprint("reservaciones_bp", __name__)

@reservaciones_bp.route("/reservaciones")
def vista_reservaciones():
    return render_template("reservaciones.html")

@reservaciones_bp.route("/api/reservaciones", methods=["GET"])
def obtener_reservaciones():
    conn = get_connection()
    cursor = conn.cursor(dictionary=True)

    query = """
        SELECT
            r.id_reservacion,
            r.no_empleado,
            r.nombre_cliente,
            r.apellido_cliente,
            r.telefono,
            r.fecha,
            r.hora,
            r.no_personas,
            r.estado,
            r.comentarios
        FROM Reservacion r
        ORDER BY r.fecha ASC, r.hora ASC
    """

    cursor.execute(query)
    reservaciones = cursor.fetchall()

    for r in reservaciones:
        if r["fecha"]:
            r["fecha"] = r["fecha"].strftime("%Y-%m-%d")

        if r["hora"]:
            total_seconds = int(r["hora"].total_seconds())
            hours = total_seconds // 3600
            minutes = (total_seconds % 3600) // 60
            seconds = total_seconds % 60
            r["hora"] = f"{hours:02d}:{minutes:02d}:{seconds:02d}"

    cursor.close()
    conn.close()

    return jsonify(reservaciones)

@reservaciones_bp.route("/api/reservaciones", methods=["POST"])
def crear_reservacion():
    data = request.get_json()

    conn = get_connection()
    cursor = conn.cursor()

    query = """
        INSERT INTO Reservacion
        (no_empleado, nombre_cliente, apellido_cliente, telefono, fecha, hora, no_personas, estado, comentarios)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
    """

    values = (
        data["no_empleado"],
        data["nombre_cliente"],
        data.get("apellido_cliente", ""),
        data["telefono"],
        data["fecha"],
        data["hora"],
        data["no_personas"],
        data["estado"],
        data.get("comentarios", "")
    )

    cursor.execute(query, values)
    conn.commit()

    nuevo_id = cursor.lastrowid

    cursor.close()
    conn.close()

    return jsonify({
        "ok": True,
        "message": "Reservación creada correctamente",
        "id_reservacion": nuevo_id
    }), 201

@reservaciones_bp.route("/api/reservaciones/<int:id_reservacion>", methods=["PUT"])
def actualizar_reservacion(id_reservacion):
    data = request.get_json()

    conn = get_connection()
    cursor = conn.cursor()

    query = """
        UPDATE Reservacion
        SET nombre_cliente=%s,
            apellido_cliente=%s,
            telefono=%s,
            fecha=%s,
            hora=%s,
            no_personas=%s,
            estado=%s,
            comentarios=%s
        WHERE id_reservacion=%s
    """

    values = (
        data["nombre_cliente"],
        data.get("apellido_cliente", ""),
        data["telefono"],
        data["fecha"],
        data["hora"],
        data["no_personas"],
        data["estado"],
        data.get("comentarios", ""),
        id_reservacion
    )

    cursor.execute(query, values)
    conn.commit()

    if cursor.rowcount == 0:
        cursor.close()
        conn.close()
        return jsonify({"ok": False, "message": "Reservación no encontrada"}), 404

    cursor.close()
    conn.close()

    return jsonify({"ok": True, "message": "Reservación actualizada correctamente"})

@reservaciones_bp.route("/api/reservaciones/<int:id_reservacion>", methods=["DELETE"])
def eliminar_reservacion(id_reservacion):
    conn = get_connection()
    cursor = conn.cursor()

    cursor.execute("DELETE FROM Reservacion WHERE id_reservacion = %s", (id_reservacion,))
    conn.commit()

    if cursor.rowcount == 0:
        cursor.close()
        conn.close()
        return jsonify({"ok": False, "message": "Reservación no encontrada"}), 404

    cursor.close()
    conn.close()

    return jsonify({"ok": True, "message": "Reservación eliminada correctamente"})